Extend cilium-operator binary to be used as command line tool

fristonio opened the pull request:
This PullRequest aims at extending the `cilium-operator` binary to be used as a command-line tool. There are two reworks associated with this PR.
1. Make `cilium-operator` to use `openapi` config as apiserver implementation similar to what we have for Cilium and `cilium-health`.
2. Add metric subcommand to `cilium-operator` to list metrics from a running cilium-operator instance apiserver.

Fixes #13132

![image](https://user-images.githubusercontent.com/21292343/103005084-dac8dc00-4557-11eb-8425-a4e7cf8f168d.png)

fristonio committed:
api: add openapi config for the API served by the Operator

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
operator: add metrics list command for cilium-operator

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
operator: add operator apiserver implementation using openapi gen code

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
operator: integrate openapi server implementation with cilium-operator

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

sayboras commented on api/v1/operator/openapi.yaml:
```diff
@@ -0,0 +1,54 @@
+  version: v1beta
+x-schemes:
+- unix
+basePath: "/v1beta"
```
Are we following the convention in which basepath is same as version? This https://github.com/cilium/cilium/blob/3b56fc99a42ad6961fdd30a69e211e5777b36d9f/api/v1/openapi.yaml#L9 seems to have user friendly `v1` for basepath only.

fristonio replied:
Since its an initial version for the operator API I wanted to keep it as `v1beta` similar to what we have for `cilium-health`. I think we can change the base path to `/v1` if everyone agrees.

fristonio replied:
Discussed offline with @aanm 
It makes sense to keep the basepath as `/v1` but the version should be `v1beta`.

sayboras commented on api/v1/operator/openapi.yaml:
```diff
@@ -0,0 +1,54 @@
+    properties:
+      status:
+        description: Status of Cilium operator
+        type: string
```
nit: new line :thinking:

sayboras commented on operator/main.go:
```diff
@@ -112,6 +114,11 @@ var (
 	isLeader atomic.Value
 )
 
+func init() {
```
I find it's a little bit confusing to have same binary for CLI + deamon server, it's sometimes confusing to me. Keen to hear inputs from others.

sayboras commented on operator/cmd/root.go:
```diff
@@ -0,0 +1,19 @@
+
+// Populate options required by cilium-operator command line only.
+func Populate() {
+	operatorAddr = viper.GetString(option.OperatorAPIServeAddr)
```
This one is interesting. Compared to cilium-agent, operator only runs in one or two nodes (in HA mode), so are we expecting end user to run `operator metrics list` in one of these nodes ? Is there any difference between active and standby node ? if yes, user might need to know the leader node to run this command.

fristonio replied:
So the metrics will be listed individually for the two instances in HA mode. For the leader operator we can have extra metrics registered like `identity_gc_*` along with the standard process metrics. The idea was to provide some sort of method within `cilium-operator` container to gather information about the operator, initially metrics.
I think what you said makes sense and we can also add a status command for the operator which can provide status information of the underlying instance. :rocket: 
I can take that up as a follow-up issue.

sayboras replied:
thanks, sounds good to me :+1:

sayboras reviewed (commented):
TIL about go-swagger generation :+1: 

Couple of comments I have as below, will do another round of review later.

aanm commented on operator/cmd/metrics_list.go:
```diff
@@ -0,0 +1,75 @@
+package cmd
```
Missing copyright header

aanm commented on operator/cmd/root.go:
```diff
@@ -0,0 +1,20 @@
+package cmd
```
Missing copyright header

aanm commented on install/kubernetes/cilium/templates/cilium-operator-deployment.yaml:
```diff
@@ -154,7 +154,7 @@ spec:
             host: '::1'
 {{- end }}
-            path: /healthz
+            path: /v1beta/healthz
```
Can we avoid changing this path?

fristonio replied:
We can set the base path for open API spec to `/` and then we can avoid this change. What do you think should  be the preferred basePath for the cilium-operator API?

aanm replied:
Can't we have an alias path to also serve `/healthz`?

fristonio replied:
One way I see is to initialize the `http.Server` using the default mux with a specific entry for `/healthz` and the default handler will fallback to open-api router.
I will check if this works.

fristonio replied:
Seems to work

![image](https://user-images.githubusercontent.com/21292343/103642467-2e382280-4f79-11eb-93a1-dfc891227a9d.png)

christarazi commented on api/v1/operator/client/cilium_operator_client.go:
```diff
@@ -0,0 +1,120 @@
+// Code generated by go-swagger; DO NOT EDIT.
```
(Noting the commit title here, not the actual change)

Commit title says "operator apiserver" which is slightly confusing to me as "apiserver" usually refers to the K8s apiserver. Could you clarify the title or simply avoid using "apiserver"? Maybe you meant "the API served by the Operator"?

fristonio replied:
Done

christarazi commented on operator/api/server.go:
```diff
@@ -0,0 +1,174 @@
+			return
+		}
+		// Allow reuse of recently-used ports. This gives the agent a
+		// better change to re-bind upon restarts.
```
This should refer to the Operator, rather than the agent?

fristonio replied:
Ah, this function was taken from the previous implementation that lived inside the operator. I can update the comment. :sweat_smile:

christarazi reviewed (changes requested):
Looks good! A few more minor comments on top of everyone else's review

christarazi reviewed (approved):
LGTM besides already pending items

kaworu reviewed (approved):
Overall LGTM. I think `CODEOWNERS` should be updated to take into account the new `/api/v1/operator` directory:

```diff
diff --git a/CODEOWNERS b/CODEOWNERS
index bdbe06c83..29d6a492e 100644
--- a/CODEOWNERS
+++ b/CODEOWNERS
@@ -35,6 +35,7 @@
 /api/ @cilium/api
 /api/v1/flow/ @cilium/api @cilium/hubble
 /api/v1/observer/ @cilium/api @cilium/hubble
+/api/v1/operator/ @cilium/api @cilium/operator
 /api/v1/peer/ @cilium/api @cilium/hubble
 /bpf/ @cilium/bpf
 Makefile* @cilium/build
```

fristonio committed:
github: update CODEOWNERS for cilium operator api

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

aanm commented on api/v1/operator/openapi.yaml:
```diff
@@ -0,0 +1,54 @@
+        '200':
+          description: Cilium operator is healthy
+          schema:
+            "$ref": "#/definitions/StatusResponse"
```
This should not have a schema.

fristonio replied:
Ok, I did not realize that we need to preserve the previously used plain text schema for `/healthz`

aanm commented on operator/api/server.go:
```diff
@@ -106,9 +112,23 @@ func (s *Server) StartServer() error {
+		// to be specified.
+		mux.HandleFunc("/healthz", func(rw http.ResponseWriter, _ *http.Request) {
+			resp := s.healthzHandler.Handle(operator.GetHealthzParams{})
+			rw.Header().Set("Content-Type", "application/json")
```
I was almost missing this, setting the header in here made me double check. The current master is simply returning `ok` with content-type `plain/text`. This healthz should not be modified in the `/healthz` nor `/v1/healthz`.

```
$ curl -vvv 127.0.0.1:9234/healthz
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 9234 (#0)
> GET /healthz HTTP/1.1
> Host: 127.0.0.1:9234
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Tue, 05 Jan 2021 16:49:13 GMT
< Content-Length: 2
< Content-Type: text/plain; charset=utf-8
< 
* Connection #0 to host 127.0.0.1 left intact
ok
```

fristonio replied:
I have pushed the required changes to keep the schema the same as before. 
![image](https://user-images.githubusercontent.com/21292343/103748600-60ee2380-502a-11eb-9e19-bf9fa0248e9b.png)

fristonio committed:
operator: add separate api handler for /healthz requests

* By default the open-api spec contains the /v1 basepath. This commit
  adds a separate handler for /healthz requests outside of the default
  basepath. This is done using HTTP server mux where the index handler
  is the default open-api router and `/healthz` requests are handled by
  a new handler.

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

sayboras reviewed (approved):
LGTM 💯
